// Contains protocol buffer messages for application architecture representation

syntax = "proto3";

option java_multiple_files = true;
option java_package = "";
option java_outer_classname = "";
option objc_class_prefix = "";

package cloud_controller;

message Architecture {
    string name = 1;
    map<string, Component> components = 2;
    Secret secret = 3;
    ApplicationType type = 4;
    string job_id = 5;
    string code = 6;
    string parameters = 7;
    string config = 8;
    int32 minimal_interval = 9;
    string docker_container = 10;
    string min_memory = 11;
    string max_memory = 12;
    string min_cpu = 13;
    string max_cpu = 14;
    string k8s_labels = 15;
    string signal_set = 16;
    string execution_time_signal = 17;
    string run_count_signal = 18;
}

message Component {
    string deployment = 1;
    repeated Probe probes = 2;
    repeated string dependsOn = 4;
    string name = 5;
    Cardinality cardinality = 6;
    string id = 7;
    ComponentType type = 9;
    Statefulness statefulness = 10;
    UEMPolicy policy = 11;
    repeated string whitelist = 12;
}

enum ApplicationType {
    REGULAR = 0;
    IVIS = 1;
}

enum Statefulness {
    NONE = 0;
    MONGO = 2;
}

enum UEMPolicy {
    DEFAULT = 0;
    FREE = 1;
    WHITELIST = 2;
}

enum Cardinality {
    MULTIPLE = 0;
    SINGLE = 1;
}

enum ComponentType {
    MANAGED = 0;
    UNMANAGED = 1;
}

message Secret {
    string value = 1;
}

//-----------------------------PROBES-------------------------------

message Probe {
    string name = 1;
    string application = 2;
    string component = 3;
    repeated Contract requirements = 5;
    string alias = 6;
}

message TimeContract {
    int32 time = 1;
    int32 percentile = 2;
}

message ThroughputContract {
    int32 requests = 1;
    int32 per = 2;
}

message Contract {
    oneof type {
        TimeContract time = 1;
        ThroughputContract throughput = 2;
    }
}

message ApplicationTimingRequirements {
    string name = 1;
    repeated TimeContract contracts = 2;
}
